File: ok_api.py
import requests
import json
import urllib.parse
from  classes.soup_profile import *

def auth(login,password):
    url = "https://api.ok.ru/api/batch/executeV2"
    payload = 'application_key=CBAFJIICABABABABA&auth.login=&methods=%5B%0A%20%20%20%20%7B%0A%20%20%20%20%20%20%20%20%22auth.login%22%3A%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%22params%22%3A%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22client%22%3A%20%22android_8_20.4.20%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22deviceId%22%3A%20%22INSTALL_ID%3D030b17c2-f51d-4641-b860-c9230a751ec5%3BANDROID_ID%3D84949fde69b77119%3B%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22gen_token%22%3A%20true%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22password%22%3A%20%22'+password+'%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22user_name%22%3A%20%22'+login+'%22%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22verification_supported%22%3A%20true%2C%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22verification_supported_v%22%3A%20%223%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%7D%0A%5D'
    headers = {
    'User-Agent': 'OKAndroid/20.4.20 b614 (Android 7.0; en_US; Android Custom Phone Build/vbox86p-userdebug 7.0 NRD90M 391 test-keys; xhdpi 320dpi 720x1184)',
    'Accept': 'application/json',
    'Content-Type': 'application/x-www-form-urlencoded',
    'Cookie': 'bci=-3218965118927310717; _statid=b877bcdf-b9bd-40d6-9715-53452ec7df84'
    }
    response = requests.request("POST", url, headers=headers, data = payload)
    res = json.loads(response.text,encoding='utf8')[0]["ok"]
    return res

def user_search(session_key,search_string,in_city=None,in_age_from=None,in_age_to=None):
    url = "https://api.ok.ru/api/batch/executeV2"
    city = ''
    age_from = ''
    age_to = ''
    if in_city is not None:
        city = ',\\"city\\":\\"'+in_city+'\\"'
    if in_age_from is not None:
        age_from=',\\"min_age\\":'+str(in_age_from)
    if in_age_to is not None:
        age_to=',\\"max_age\\":'+str(in_age_to)     
    method = '[{"search.globalGrouped":{"params":{"count":100,"fieldset":"android.1","filters":"[{\\"type\\":\\"user\\"'+city+age_from+age_to+'}]","query":"'+search_string+'","screen":"GLOBAL_SEARCH_USERS_NO_RESULTS","types":"USER"}}}]'
    method = urllib.parse.quote_plus(method)
    payload = '389%0A__screen=feed_main&application_key=CBAFJIICABABABABA&methods='+method+'&session_key='+session_key            
    headers = {
    'Content-Type': 'application/x-www-form-urlencoded',
    'User-Agent': 'OKAndroid/20.5.15 b619 (Android 7.1.2; en_US; samsung SM-G955N Build/N2G48H.G9550ZHU1AQEE; hdpi 240dpi 720x1280)',
    'Accept': 'application/json',    
    'Host': 'api.ok.ru',
    'Connection': 'Keep-Alive',
    'Accept-Encoding': 'gzip',
    'Cookie': 'bci=-3218965118927310717; _statid=b877bcdf-b9bd-40d6-9715-53452ec7df84'
    }
    response = requests.request("POST", url, headers=headers, data = payload)
    if response.status_code!=200:
        return None
    res = json.loads(response.text,encoding='utf8')[0]["ok"]
    if "entities" not in res:
        return None
    res = res["entities"]["users"]
    method = '[{"users.getInfo":{"params":{"fields":"URL_PROFILE,name,last_name,first_name,pic190x190,last_online_ms,birthday,gender,age,location","uids":"'
    profile_num = 0
    profile_max = len(res)
    for profile in res:        
        method+=profile["uid"]
        if profile_num<profile_max-1:
            method+=","        
        profile_num+=1
    method+='"}}}]'
    method = urllib.parse.quote_plus(method)
    payload = '389%0A__screen=feed_main&application_key=CBAFJIICABABABABA&methods='+method+'&session_key='+session_key
    response = requests.request("POST", url, headers=headers, data = payload)
    if response.status_code!=200:
        return None
    response_text = response.text
    # response_text='[{"ok": {"user": {"uid": "296225027453", "birthday": "1981-08-13", "birthdaySet": true, "age": 39, "first_name": "виталий", "last_name": "принцев", "name": "виталий принцев", "gender": "male", "location": {"city": "армянск", "country": "UKRAINE", "countryCode": "UA", "countryName": "Украина"}, "pic190x190": "https://api.ok.ru/img/stub/user/male/190.png", "url_profile": "https://ok.ru/profile/520647022216"}}}, {"ok": {"user": {"uid": "781313734737", "birthday": "1964-03-08", "birthdaySet": true, "age": 56, "first_name": "Виталий", "last_name": "Принцев", "name": "Виталий Принцев", "gender": "male", "location": {"city": "", "country": "LITHUANIA", "countryCode": "LT", "countryName": "Литва"}, "last_online_ms": 1596024928396, "pic190x190": "https://i.mycdn.me/image?id=896535025572&t=33&plc=API&aid=25662464&tkn=*7IxTAQcbwLTnEukrGKQUBsDsbKA", "pic_base": "https://i.mycdn.me/image?id=896535025572&plc=API&aid=25662464&tkn=*9ZBEb26SwsxsndSb51gppJoSjKY", "url_profile": "https://ok.ru/profile/584811109284"}}}, {"ok": {"user": {"uid": "790953514277", "birthday": "1964-03-08", "birthdaySet": true, "age": 56, "first_name": "Виталий", "last_name": "Принцев", "name": "Виталий Принцев", "gender": "male", "location": {"city": "", "country": "KAZAKHSTAN", "countryCode": "KZ", "countryName": "Казахстан"}, "last_online_ms": 1562841392875, "pic190x190": "https://i.mycdn.me/image?id=868017411536&t=33&plc=API&aid=25662464&tkn=*Dw7OtCv_ymoknRxn5gs-_DDsF1Q", "pic_base": "https://i.mycdn.me/image?id=868017411536&plc=API&aid=25662464&tkn=*i61hMiEdGoaiGQsZJOCzic6JJIA", "url_profile": "https://ok.ru/profile/575124600528"}}}, {"ok": {"user": {"uid": "489604452981", "birthday": "1960-09-04", "birthdaySet": true, "age": 60, "first_name": "Виталий", "last_name": "Принцев", "name": "Виталий Принцев", "gender": "male", "location": {"city": "Нея", "country": "RUSSIAN_FEDERATION", "countryCode": "RU", "countryName": "Россия"}, "pic190x190": "https://api.ok.ru/img/stub/user/male/190.png", "url_profile": "https://ok.ru/profile/327460532608"}}}, {"ok": {"user": {"uid": "790581113250", "birthday": "1976-06-06", "birthdaySet": true, "age": 44, "first_name": "Виталий", "last_name": "Принцев", "name": "Виталий Принцев", "gender": "male", "location": {"city": "", "country": "RUSSIAN_FEDERATION", "countryCode": "RU", "countryName": "Россия"}, "last_online_ms": 1593185912654, "pic190x190": "https://i.mycdn.me/image?id=893997786711&t=33&plc=API&ts=0000000000000002e9&aid=25662464&tkn=*dChYBdYo-2P_1YjmshAaTqpkJFY", "pic_base": "https://i.mycdn.me/image?id=893997786711&ts=0000000000000002e9&plc=API&aid=25662464&tkn=*2pFbIUpTLeUBJcttPAqpwYsggUQ", "url_profile": "https://ok.ru/profile/574767911511"}}}, {"ok": {"user": {"uid": "484251374420", "birthday": "1960-09-04", "birthdaySet": true, "age": 60, "first_name": "Виталий", "last_name": "Принцев", "name": "Виталий Принцев", "gender": "male", "location": {"city": "Нея", "country": "RUSSIAN_FEDERATION", "countryCode": "RU", "countryName": "Россия"}, "last_online_ms": 1354118904770, "pic190x190": "https://i.mycdn.me/image?id=244905868961&t=33&plc=API&aid=25662464&tkn=*lVTxatpFh2mex4MDNgj2uOqtDa0", "pic_base": "https://i.mycdn.me/image?id=244905868961&plc=API&aid=25662464&tkn=*fWxYshPJkbmzVbqgn74EGaCpmng", "url_profile": "https://ok.ru/profile/332846968993"}}}, {"ok": {"user": {"uid": "779339494608", "birthday": "1964-03-08", "birthdaySet": true, "age": 56, "first_name": "Виталий", "last_name": "Принцев", "name": "Виталий Принцев", "gender": "male", "location": {"city": "", "country": "KAZAKHSTAN", "countryCode": "KZ", "countryName": "Казахстан"}, "last_online_ms": 1519628024691, "pic190x190": "https://i.mycdn.me/image?id=838633079077&t=33&plc=API&aid=25662464&tkn=*6UF8GHyCegbYbCVLL1AiG_5kVnU", "pic_base": "https://i.mycdn.me/image?id=838633079077&plc=API&aid=25662464&tkn=*iK4w2HrcncYZHpEDAWi1L_evTMc", "url_profile": "https://ok.ru/profile/587149661989"}}}]'
    profiles_ok = json.loads(response_text,encoding='utf8')[0]["ok"]
    profiles_soup = []
    for profile_ok in profiles_ok:
        descriptions = []
        profile_pic = ""
        if "pic190x190" in profile_ok:
            profile_pic = profile_ok["pic190x190"]
        if "age" in profile_ok:
            descriptions.append({"age":profile_ok["age"]})
        if "birthday" in profile_ok:
            descriptions.append({"birthday":profile_ok["birthday"]})
        if "location" in profile_ok:
            location = profile_ok["location"]
            if "city" in location:
                descriptions.append({"city":location["city"]})
            if "country" in location:
                descriptions.append({"countrys":location["country"]})
        tmp_soup_profile = SOUP_Profile(full_name=profile_ok["name"],profile_url=profile_ok["url_profile"],profile_pic=profile_pic,profile_descriptions=descriptions)
        profiles_soup.append(tmp_soup_profile)
    return profiles_soup

File: ok_tools/okapi.js
const request = require('request');

var get_auth = async function(login, password) {
    var options = {
        'method': 'POST',
        'url': 'https://api.ok.ru/api/batch/executeV2',
        'headers': {
            'User-Agent': 'OKAndroid/20.4.20 b614 (Android 7.0; en_US; Android Custom Phone Build/vbox86p-userdebug 7.0 NRD90M 391 test-keys; xhdpi 320dpi 720x1184)',
            'Accept': 'application/json',
            'Transfer-Encoding': 'chunked',
            'Content-Type': 'application/x-www-form-urlencoded'
        },
        form: {
            'application_key': 'CBAFJIICABABABABA',
            'auth.login': '',
            'methods': `[\n    {\n        "auth.login": {\n            "params": {\n                "client": "android_8_20.4.20",\n                "deviceId": "INSTALL_ID=030b17c2-f51d-4641-b860-c9230a751ec5;ANDROID_ID=84949fde69b77119;",\n                "gen_token": true,\n                "password": "${password}",\n                "user_name": "${login}",\n                "verification_supported": true,\n                "verification_supported_v": "3"\n            }\n        }\n    }\n]`
        }
    };
    return new Promise((resolve, reject) => {
        request(options, function(error, response) {
            if (error) throw new Error(error);
            resolve(JSON.parse(response.body));
        });
    });
}

var get_friends = async function(uid, session_key) {
    var options = {
        'method': 'GET',
        'url': `https://api.ok.ru/api/friends/get?application_key=CBAFJIICABABABABA&client=android_8_16.9.27&fid=${uid}&session_key=${session_key}`,
        'headers': {
            'Accept': 'application/json',
            'User-Agent': 'OKAndroid/16.9.27 b231 (Android 7.0; en_US; Android Custom Build/vbox86p-userdebug; xhdpi 320dpi 720x1184)',
            'Host': 'api.ok.ru',
            'Connection': 'Keep-Alive',
            'Cookie': 'bci=-3218965118927310717; _statid=b877bcdf-b9bd-40d6-9715-53452ec7df84'
        }
    };
    return new Promise((resolve, reject) => {
        request(options, function(error, response) {
            if (error) throw new Error(error);
            resolve(JSON.parse(response.body));
        });
    });
}

var get_users_info = async function(uids, session_key) {
    var uid_str = '';
    if (uids != undefined) {
        for (var i = 0; i < uids.length; i++) {
            uid_str += uids[i];
            if (i < uids.length - 1)
                uid_str += '%2C';
        }
    }
    var options = {
        'method': 'GET',
        'url': `https://api.ok.ru/api/users/getInfo?application_key=CBAFJIICABABABABA&client=android_8_16.9.27&emptyPictures=false&fields=last_name%2Cname%2Cfirst_name%2Cpic190x190%2Conline%2Clast_online_ms%2Cpic190x190%2Cpic_full%2Cshow_lock%2Cbirthday%2Cgender%2Ccan_vcall%2Cprivate%2Ccan_vmail&session_key=${session_key}&uids=${uid_str}`,
        'headers': {
            'Accept': 'application/json',
            'User-Agent': 'OKAndroid/16.9.27 b231 (Android 7.0; en_US; Android Custom Build/vbox86p-userdebug; xhdpi 320dpi 720x1184)',
            'Host': 'api.ok.ru',
            'Connection': 'Keep-Alive',
            'Cookie': 'bci=-3218965118927310717; _statid=b877bcdf-b9bd-40d6-9715-53452ec7df84'
        }
    };
    return new Promise((resolve, reject) => {
        request(options, function(error, response) {
            if (error) throw new Error(error);
            resolve(JSON.parse(response.body));
        });
    });
}



var get_accounts_by_phones = async function(query, session_key) {
    //{"credentials":[{"firstName":"Adriana","lastName":"Poetter","phone":" 992938808418","hasAvatar":false,"hasRingtone":false,"isFavorite":false}]}
    var options = {
        'method': 'POST',
        'url': 'https://api.ok.ru/api/search/byContactsBook',
        'headers': {
            'User-Agent': 'OKAndroid/20.4.20 b614 (Android 7.0; en_US; Android Custom Phone Build/vbox86p-userdebug 7.0 NRD90M 391 test-keys; xhdpi 320dpi 720x1184)',
            'Accept': 'application/json',
            'Transfer-Encoding': 'chunked',
            'Content-Type': 'application/x-www-form-urlencoded'
        },
        form: {
            '__screen': 'friends,feed_main',
            'application_key': 'CBAFJIICABABABABA',
            'fields': 'user.url_profile,user.age,user.last_name,user.pic600x600,user.pic128x128,user.gender,user.location,user.uid,user.first_name,user.online,birthday',
            'session_key': `${session_key}`,
            'query': `${query}`
        }
    };
    return new Promise((resolve, reject) => {
        request(options, function(error, response) {
            if (error) throw new Error(error);
            resolve(JSON.parse(response.body));
        });
    });
}

var get_photo_urls = async function(session_key, uid, max_count) {
    var options = {
        'method': 'POST',
        'url': 'https://api.ok.ru/api/batch/executeV2',
        'headers': {
            'User-Agent': 'OKAndroid/20.4.20 b614 (Android 7.0; en_US; Android Custom Phone Build/vbox86p-userdebug 7.0 NRD90M 391 test-keys; xhdpi 320dpi 720x1184)',
            'Accept': 'application/json',
            'Transfer-Encoding': 'chunked',
            'Content-Type': 'application/x-www-form-urlencoded'
        },
        form: {
            'application_key': 'CBAFJIICABABABABA',
            'session_key': session_key,
            'methods': `[{"photos.getPhotos":{"params":{"aid":"stream","count":${max_count},"detectTotalCount":"true","fid":"${uid}","fields":"photo.pic_max"}}}]`
        }
    };
    return new Promise((resolve, reject) => {
        request(options, function(error, response) {
            if (error) throw new Error(error);
            resolve(JSON.parse(response.body));
        });
    });
}

var get_users_status = async function(session_key, uids) {
    var uids_str = '';
    for (var i = 0; i < uids.length; i++) {
        uids_str += uids[i];
        if (i < uids.length - 1)
            uids_str += ',';
    }
    var options = {
        'method': 'POST',
        'url': 'https://api.ok.ru/api/batch/executeV2',
        'headers': {
            'User-Agent': 'OKAndroid/20.4.20 b614 (Android 7.0; en_US; Android Custom Phone Build/vbox86p-userdebug 7.0 NRD90M 391 test-keys; xhdpi 320dpi 720x1184)',
            'Accept': 'application/json',
            'Transfer-Encoding': 'chunked',
            'Content-Type': 'application/x-www-form-urlencoded'
        },
        form: {
            'application_key': 'CBAFJIICABABABABA',
            'session_key': session_key,
            'methods': `[{"users.getInfo": {"params": {"fields": "name,last_online","uids": "${uids_str}"}}}]`
        }
    };
    return new Promise((resolve, reject) => {
        request(options, function(error, response) {
            if (error) throw new Error(error);
            resolve(JSON.parse(response.body));
        });
    });
}


var get_recovery_phone = async function(profile_url_id) {
    var options = {
        'method': 'POST',
        'url': 'https://ok.ru/dk',
        'headers': {
            'authority': 'ok.ru',
            'accept': '*/*',
            'x-requested-with': 'XMLHttpRequest',
            'content-type': 'application/x-www-form-urlencoded; charset=UTF-8',
            'strd': 'false',
            'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.105 Safari/537.36',
            'tkn': 'undefined',
            'origin': 'https://ok.ru',
            'sec-fetch-site': 'same-origin',
            'sec-fetch-mode': 'cors',
            'sec-fetch-dest': 'empty',
            'referer': 'https://ok.ru/',
            'accept-language': 'en-GB,en;q=0.9,en-US;q=0.8,ru;q=0.7',
            'cookie': 'viewport=1057; _lastSmsForCodeTime_password=; bci=-3218965118927310717; _statid=b877bcdf-b9bd-40d6-9715-53452ec7df84; JSESSIONID=98be89eea4f1e4db160c664be815496f2946e88abf1c5cef.bef3488'
        },
        body: "st.recoveryData=" + profile_url_id + "&st.ccode=&st.recoveryMethod=Link&st.registrationAction=ValidateLink&st.cmd=anonymPasswordRecoveryNew&cmd=AnonymPasswordRecoveryNew&gwt.requested=da69c7ebT1598879267705"

    };
    return new Promise((resolve, reject) => {
        request(options, function(error, response) {
            if (error) throw new Error(error);
            var JSESSIONID = "JSESSIONID=98be89eea4f1e4db160c664be815496f2946e88abf1c5cef.bef3488";
            if (response.headers["set-cookie"] != undefined) {
                response.headers["set-cookie"].forEach(cookie => {
                    if (cookie.indexOf("JSESSIONID") >= 0) {
                        JSESSIONID = cookie;
                    }
                });
            }
            var options = {
                'method': 'GET',
                'url': 'https://ok.ru/dk?st.cmd=anonymPasswordRecoveryNew&st.registrationAction=ChooseCodeDestination&st.recoveryMethod=Link',
                'headers': {
                    'authority': 'ok.ru',
                    'upgrade-insecure-requests': '1',
                    'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.105 Safari/537.36',
                    'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
                    'sec-fetch-site': 'same-origin',
                    'sec-fetch-mode': 'navigate',
                    'sec-fetch-user': '?1',
                    'sec-fetch-dest': 'document',
                    'referer': 'https://ok.ru/',
                    'accept-language': 'en-GB,en;q=0.9,en-US;q=0.8,ru;q=0.7',
                    'cookie': `viewport=1057; _lastSmsForCodeTime_password=; bci=-3218965118927310717; _statid=b877bcdf-b9bd-40d6-9715-53452ec7df84; ${JSESSIONID}`
                }
            };
            request(options, function(error, response) {
                if (error) throw new Error(error);
                var body = response.body;
                var phrase = "<div>Via SMS +";
                var phrase2 = "****";
                var phrase_ind = body.indexOf(phrase);
                if (phrase_ind < 0) {
                    resolve(false);
                }
                var body_part1 = body.substring(phrase_ind + phrase.length)
                var phrase2_ind = body_part1.indexOf(phrase2);
                if (phrase2_ind < 0) {
                    resolve(false);
                }
                var number = body_part1.substring(0, phrase2_ind);
                resolve(number);
            });
        });
    });
}



exports.get_auth = get_auth;
exports.get_users_info = get_users_info;
exports.get_friends = get_friends;
exports.get_accounts_by_phones = get_accounts_by_phones;
exports.get_photo_urls = get_photo_urls;
exports.get_users_status = get_users_status;
exports.get_recovery_phone = get_recovery_phone;